icefoganalytics/internal-data-portal
# Automate Dataset Field Creation From API Link

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/37

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/issues/35
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- https://github.com/icefoganalytics/internal-data-portal/issues/7

# Context

**Is your feature request related to a problem? Please describe.**
To speed up the generation for content for the dataset fields tab, it would be nice to be able to auto-fill as much as possible from the provided API link.

**Describe the solution you'd like**
When the user confirms the "add api link" action on the dataset description manage tab, it fetches some data from the corresponding API, and using the fetched data, builds out the list of fields that it expects.

This will likely also required that an API key is provided in the "add api link" dialog. This may or may not be the same API key that is provided to the end user to permit them to access said API.

**Describe alternatives you've considered**
Do nothing and require the user to fill in everything manually.
When adding fields to a dataset from /datasets/:slug/fields/managed page, provide the list of currently un-added fields in the field name box. Let the user select from this list when adding fields.

**Additional context**

Add api link - http://localhost:8080/datasets/test-real-user-groups/description/manage
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/18f445cd-0747-4d18-8f94-2a250610a997)

Potential Field automation - http://localhost:8080/datasets/test-real-user-groups/fields/manage
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/e811351a-b4e4-4b3c-9f89-712f6186ee92)

# Implementation

1. Move data format normalization into the JMESPath parsing service to avoid having to handle it in multiple places downstream. Also this way the parsedJsonData type is valid, and simple.
2. Add bulk replace service for dataset fields, and only trigger it if no fields exist. I went for a bulk replaces, as it might make sense to provide that as an option through the UI. It might also make sense to have a bulk upsert, that adds any new fields, and flags missing or renamed fields.
3. Change default data key to "data" as this is more human friendly that "value"

# Screenshots

Fields tab
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/bf2bbb0c-dda9-4e86-b92c-338da368de44)

Visualize tab
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/5fe818c4-51e5-4778-bd95-b8d272067682)


# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
4. Log in to the app at http://localhost:8080
5. Go to the /datesets page and create a dataset.
6. Click the "Manage API Link" button.
7. Fill in the Fields, and click create. e.g. https://api.gov.yk.ca/directory/employees
8. Pick a JMESPath transform if needed and import the dataset.
9. Go to the "Fields" tab and see all the new field import goodness.
10. Go to the Visualize tab and check that you can see the imported data.
